src/main/java/com/direwolf20/buildinggadgets/common/schema/template/TemplateData.java

```diff
@@ -0,0 +1,127 @@
+package com.direwolf20.buildinggadgets.common.schema.template;
+
+import com.google.common.base.MoreObjects;
+import net.minecraft.block.BlockState;
+import net.minecraft.nbt.CompoundNBT;
+import net.minecraft.util.Mirror;
+import net.minecraft.util.Rotation;
+import net.minecraft.util.math.BlockPos;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * This class is a mutable wrapper around {@link BlockPos position} and {@link BlockState state} it therefore represents
+ * a single data point in a {@link Template}.
+ */
+@Immutable
+public final class TemplateData {
+    private final BlockPos pos;
+    private final BlockData data;
+
+    public TemplateData(BlockPos pos, BlockData data) {
+        this.pos = pos.toImmutable();
+        this.data = data;
+    }
+
+    public BlockPos getPos() {
+        return pos;
+    }
+
+    public BlockData getData() {
+        return data;
+    }
+
+    public TemplateData copy() {
+        return new TemplateData(pos, data.copy());
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (! (o instanceof TemplateData)) return false;
+
+        final TemplateData that = (TemplateData) o;
+
+        if (! pos.equals(that.pos)) return false;
+        return data.equals(that.data);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = pos.hashCode();
+        return 31 * result + data.hashCode();
+    }
+
+    @Override
+    public String toString() {
+        return MoreObjects.toStringHelper(this)
+                .add("pos", pos)
+                .add("data", data)
+                .toString();
+    }
+
+    @Immutable
+    public static final class BlockData {
```

Comment:
@MajorTuvok whats the point in BlockData, it seems to me you have just copied BlockState, also how come such a big Class is nested within another it should be its own class within its own file to me, just to keep files brief and easier to read.

Reply:
As you can clearly see I have not copied blockstate. I need something to aggreagate the whole data a block has: a BlockState and potentially a CompoundNBT for tile data. (see Template.Builder). Furthermore I don't like making this a seperate class, as it belongs to TemplateData, this makes the relationship more clear.

Reply:
Also this is not a big class?!?

Reply:
Wouldn't a pair be better rather than adding 100+ more lines to match a BlockState to a possible CompoundNBT.

Reply:
`Pair<BlockState, Optional<CompundNbt>>`

Reply:
No. In general one should avoid Pair's in non-local code as they are very unexpressive. I would have to add the fields to TemplateData anyway and therefore wouldn't win anything, if I converted my internal code. Not going to do this.

Reply:
> Pair<BlockState, Optional<CompundNbt>>
Optional's should be avoided in stored fields - they are meant for intermediate computations.

Reply:
Also what are the methods for as i can't currently see them being used anywhere.

Reply:
mirror and rotate are used in Template#mirror and Template#rotate respectively. copyNBT is for use when building it and getNBT for serialization (compare the comment).

Reply:
> I would have to add the fields to TemplateData

If both of you really reallly want me to, I could do this and use a `Pair<BlockState,CompoundNBT>` in Template#Builder. I however dislike this approach, as it is less expressive.

Reply:
I also use BlockData in some other intermediate computations where I would have to replace that...